### PR TITLE
Change setup_baas.rb to Swift; Fix testAsyncOpenTimeout

### PR DESCRIPTION
**Pull request by jsflax**

### CODE REVIEW

File: Realm/ObjectServerTests/RealmServer.swift [hunk 1]
```diff
@@ -49,6 +55,282 @@ extension URLSession {
...
+        process.environment = environment
+        createProcesses(commandString, process: &process, processes: &processes)
+
+        processes.forEach {
+            $0.launch()
+        }
+        processes.forEach {
+            $0.waitUntilExit()
+        }
+        if let _ = (processes.last!.standardOutput as? Pipe)?.fileHandleForReading.readDataToEndOfFile() {
```
Comment (leemaguire) [inline Realm/ObjectServerTests/RealmServer.swift:166]: Is there a reason this is written twice, and has an unused block?

Comment (jsflax) [reply Realm/ObjectServerTests/RealmServer.swift:166]: The unused block is some leftover cruft, will remove

File: Realm/ObjectServerTests/RealmServer.swift [hunk 2]
```diff
@@ -49,6 +55,282 @@ extension URLSession {
...
+
+        let stitchWorktree = goRoot
+            .appendingPathComponent("src")
+            .appendingPathComponent("github.com")
+            .appendingPathComponent("10gen")
+            .appendingPathComponent("stitch")
+
+
+        if !FileManager.default.fileExists(atPath: stitchDir.appendingPathComponent(".git").absoluteString) {
+
```
Comment (leemaguire) [inline Realm/ObjectServerTests/RealmServer.swift:233]: extra newline

**Review by pavel-ship-it — APPROVED** [review]
LGTM
less ruby for best

File: Realm/ObjectServerTests/RealmServer.swift [hunk 3]
```diff
@@ -736,7 +1059,19 @@ public class RealmServer: NSObject {
...
+        ]
+
+        let rules = app.services[serviceId].rules
+
+        rules.post(on: group, dogRule, failOnError)
+        rules.post(on: group, personRule, failOnError)
+        rules.post(on: group, hugeSyncObjectRule, failOnError)
+        // When running ObjcObjectServerTests,
+        // we do not want to pull in swift schema reqs
+        #if SWIFT_PACKAGE && REALM_HAVE_COMBINE
```
Comment (leemaguire) [inline Realm/ObjectServerTests/RealmServer.swift:1071]: Wont using these macros mean this class will cause issues when running tests from the xcproj?

**Review by leemaguire — APPROVED** [review]
LGTM, one comment

File: Realm/ObjectServerTests/RealmServer.swift [hunk 4]
```diff
@@ -20,6 +20,12 @@ import Foundation
 import RealmSwift
 import XCTest
 
+private extension URL {
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RealmServer.swift:23]: This looks like an incorrect workaround for the source URL being constructed incorrectly.

File: Realm/ObjectServerTests/RealmServer.swift [hunk 5]
```diff
@@ -285,42 +375,84 @@ public class RealmServer: NSObject {
...
-    /// Log level for the server and mongo processes.
-    public var logLevel = LogLevel.none
-
-    /// Process that runs the local mongo server. Should be terminated on exit.
-    private let mongoProcess = Process()
-    /// Process that runs the local backend server. Should be terminated on exit.
-    private let serverProcess = Process()
+        do {
+            return try JSONDecoder().decode(Dependencies.self,
+                                            from: JSONEncoder().encode(dependenciesString.components(separatedBy: "\n").dropLast().reduce(into: [String: String]()) {
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RealmServer.swift:401]: This line really needs to be split into something more reasonable.

File: Realm/ObjectServerTests/RealmServer.swift [hunk 6]
```diff
@@ -341,11 +474,187 @@ public class RealmServer: NSObject {
...
+            shutdownProcess.waitUntilExit()
+            mongoProcess.terminate()
+        }
+    }
+
+    private func setupStitch() throws {
+        print("Setting up stitch")
+
+        let goRoot = buildDir.appendingPathComponent("go")
+        if try fileManager.contentsOfDirectory(atPath: binDir.absoluteString).count == 7 {
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RealmServer.swift:514]: This is a really brittle way to do this.

File: Realm/ObjectServerTests/SwiftObjectServerTests.swift [hunk 7]
```diff
@@ -27,24 +27,10 @@ import RealmSyncTestSupport
...
-        let fakeDataSize = 1000000
-        return SwiftHugeSyncObject(value: ["data": Data(repeating: 16, count: fakeDataSize)])
-    }
-}
-
 extension User {
     func configuration(testName: String) -> Realm.Configuration {
         var config = self.configuration(partitionValue: testName)
-        config.objectTypes = [SwiftPerson.self, SwiftHugeSyncObject.self]
+        config.objectTypes = [SwiftPerson.self, Person.self, Dog.self, HugeSyncObject.self, SwiftHugeSyncObject.self]
```
Comment (tgoyne) [inline Realm/ObjectServerTests/SwiftObjectServerTests.swift:33]: These tests shouldn't need the obj-c types.

File: Realm/ObjectServerTests/RealmServer.swift [hunk 8]
```diff
@@ -341,11 +474,187 @@ public class RealmServer: NSObject {
...
+            print("server binary built")
+        }
+
+        subprocess.popen("/bin/cp stitch/etc/configs/test_config.json ./")
+    }
+
+    private func buildServer() throws {
+        try fileManager.createDirectory(at: tempDir, withIntermediateDirectories: true)
+        if !fileManager.fileExists(atPath: buildDir.absoluteString) {
+            try! FileManager.default.createDirectory(at: URL.init(fileURLWithPath: buildDir.absoluteString,
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RealmServer.swift:635]: Redundant .init

File: RealmSwift/Tests/CombineTests.swift [hunk 9]
```diff
@@ -842,14 +842,16 @@ private class CombineCollectionPublisherTests<Collection: RealmCollection>: Comb
...
         cancellable = collection
             .changesetPublisher
             .saveToken(on: self, at: \.notificationToken)
             .sink { _ in
                 calls += 1
+                ex.fulfill()
             }
         XCTAssertNotNil(notificationToken)
-
+        waitForExpectations(timeout: 2.0, handler: nil)
```
Comment (tgoyne) [inline RealmSwift/Tests/CombineTests.swift:854]: This wait shouldn't be needed and if it does fix the test then it's just working around a bug which the test was exposing.

Comment (jsflax) [reply RealmSwift/Tests/CombineTests.swift:854]: The test always passes for me locally, this was to test on the CI.

Why isn't this wait needed though? Isn't the initial `sink` value still called asynchronously?

Comment (tgoyne) [reply RealmSwift/Tests/CombineTests.swift:854]: I think maybe rather than this wait being unnecessary, there's actually two different things that should be tested. Adding a notifier and then immediately skipping it before receiving the initial notification is something that should work. This is what the test was previously testing on CI, but for that it needs a wait *after* the writes because it's possible for all of them to happen before initial is ever delivered. Adding the wait here validates that it successfully skips after receiving the initial notification, which of course it also something that should be tested.

So basically the test is correct, but there should be another copy of this test added with the wait after the loop but before the assert on call count.

File: Realm/ObjectServerTests/RLMSyncTestCase.mm [hunk 10]
```diff
@@ -451,6 +452,7 @@ - (void)setUp {
     [NSFileManager.defaultManager removeItemAtURL:self.clientDataRoot error:nil];
     [NSFileManager.defaultManager createDirectoryAtURL:self.clientDataRoot
                            withIntermediateDirectories:YES attributes:nil error:nil];
+//    [[RealmServer shared] dropTestData];
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RLMSyncTestCase.mm:455]: commented out code

File: Realm/ObjectServerTests/RealmServer.swift [hunk 11]
```diff
@@ -684,6 +989,34 @@ public class RealmServer: NSObject {
...
 
+        let swiftHugeSyncObjectRule: [String: Any] = [
+            "database": "test_data",
+            "collection": "SwiftHugeSyncObject",
+            "roles": [[
+                "name": "default",
+                        "apply_when": [:],
+                "insert": true,
+                "delete": true,
+                        "additional_fields": [:]
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RealmServer.swift:1000]: indentation is weird

File: build.sh [hunk 12]
```diff
@@ -1106,8 +1101,7 @@ case "$COMMAND" in
 
             if [[ "$target" = *"server"* ]] || [[ "$target" = "swiftpm"* ]]; then
                 source "$(brew --prefix nvm)/nvm.sh" --no-use
-                nvm install 8.11.2
-                sh build.sh setup-baas
+                nvm install 14.15.4
```
Comment (tgoyne) [inline build.sh:1104]: Grab node version from dependencies.list?

**Comment by bmunkholm** [comment]
@jsflax Is this still relevant?

**Comment by jsflax** [comment]
We could probably still use this but it was never necessary.
